alvr/client_openxr/src/extra_extensions/spatial_marker_tracking.rs

```diff
@@ -0,0 +1,299 @@
+
+        if self.snapshot_future.is_none()
+            && self.discovery_enabled.value()
+            && new > self.discovery_timeout_deadline
+        {
+            let snapshot_create_info = sys::SpatialDiscoverySnapshotCreateInfoEXT {
+                ty: sys::SpatialDiscoverySnapshotCreateInfoEXT::TYPE,
+                next: ptr::null(),
+                component_type_count: self.enabled_components.len() as u32,
+                component_types: self.enabled_components.as_ptr(),
+            };
+
+            let mut create_snapshot_future: sys::FutureEXT = 0;
+            unsafe {
+                super::xr_res((self
+                    .spatial_entity_fns
+                    .create_spatial_discovery_snapshot_async)(
+                    self.inner,
+                    &snapshot_create_info,
+                    &mut create_snapshot_future,
+                ))?;
+            }
+
+            self.snapshot_future = Some(create_snapshot_future);
+            self.discovery_timeout_deadline = new + DISCOVERY_TIMEOUT;
+        };
+
+        let &Some(future) = &self.snapshot_future else {
+            return Ok(None);
+        };
+
+        // Return if the future is not completed
+        if !super::check_future(&self.instance, future)? {
+            return Ok(None);
+        }
+
+        self.snapshot_future = None;
+
+        let completion_info = sys::CreateSpatialDiscoverySnapshotCompletionInfoEXT {
+            ty: sys::CreateSpatialDiscoverySnapshotCompletionInfoEXT::TYPE,
+            next: ptr::null(),
+            base_space: base_space.as_handle(),
+            time,
+            future,
+        };
+
+        let completion = unsafe {
+            let mut completion =
+                sys::CreateSpatialDiscoverySnapshotCompletionEXT::out(ptr::null_mut());
+            super::xr_res((self
+                .spatial_entity_fns
+                .create_spatial_discovery_snapshot_complete)(
+                self.inner,
+                &completion_info,
+                completion.as_mut_ptr(),
+            ))?;
+            completion.assume_init()
+        };
+        if completion.future_result != sys::Result::SUCCESS {
+            return Err(completion.future_result);
+        }
+
+        let query_contition = sys::SpatialComponentDataQueryConditionEXT {
+            ty: sys::SpatialComponentDataQueryConditionEXT::TYPE,
+            next: ptr::null(),
+            component_type_count: self.enabled_components.len() as u32,
+            component_types: self.enabled_components.as_ptr(),
+        };
+
+        let mut query_result = sys::SpatialComponentDataQueryResultEXT {
+            ty: sys::SpatialComponentDataQueryResultEXT::TYPE,
+            next: ptr::null_mut(),
+            entity_id_capacity_input: MAX_MARKERS_COUNT as u32,
+            entity_id_count_output: 0,
+            entity_ids: self.entity_ids.as_mut_ptr(),
+            entity_state_capacity_input: MAX_MARKERS_COUNT as u32,
```

Comment:
You should still be using the size queried from the arrays themselves, to ensure that possible changes don't mess things up.